Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/NamespaceSynchronizationConfig.swift

```diff
@@ -0,0 +1,184 @@
+            syncedDocuments: try! docsColl
+                .find(CoreDocumentSynchronization.filter(forNamespace: namespace))
+                .reduce(into: [HashableBSONValue: CoreDocumentSynchronization.Config](), { (syncedDocuments, config) in
+                    syncedDocuments[config.documentId] = config
+                }))
+
+    }
+
+    init(namespacesColl: MongoCollection<NamespaceSynchronization.Config>,
+         docsColl: MongoCollection<CoreDocumentSynchronization.Config>,
+         config: inout Config) {
+        self.namespacesColl = namespacesColl
+        self.docsColl = docsColl
+        self.config = config
+    }
+
+    /// Make an iterator that will iterate over the associated documents.
+    func makeIterator() -> NamespaceSynchronization.Iterator {
+        return NamespaceSynchronizationIterator.init(docsColl: docsColl,
+                                                     values: config.syncedDocuments.values)
+    }
+
+    /**
+     Read or set a document configuration from this namespace.
+     If the document does not exist, one will not be returned.
+     If the document does exist and nil is supplied, the document
+     will be removed.
+     - parameter documentId: the id of the document to read
+     - returns: a new or existing NamespaceConfiguration
+     */
+    subscript(documentId: HashableBSONValue) -> CoreDocumentSynchronization? {
+        get {
+            nsLock.readLock()
+            defer { nsLock.unlock() }
+            guard var config = config.syncedDocuments[documentId] else {
+                return nil
+            }
+            return CoreDocumentSynchronization.init(docsColl: docsColl, config: &config)
+        }
+        set(value) {
+            nsLock.writeLock()
+            defer { nsLock.unlock() }
+            
+            guard let value = value else {
+                try! docsColl.deleteOne(docFilter(forNamespace: config.namespace,
+                                                  withDocumentId: documentId.bsonValue))
+                config.syncedDocuments[documentId] = nil
+                return
+            }
+
+            try! docsColl.replaceOne(
+                filter: docFilter(forNamespace: self.config.namespace,
+                                  withDocumentId: documentId.bsonValue),
+                replacement: value.config,
+                options: ReplaceOptions.init(upsert: true))
+            self.config.syncedDocuments[documentId] = value.config
+        }
+    }
+
+    /**
+     Configure a ConflictHandler and ChangeEventListener to this namespace.
+     These will be used to handle conflicts or listen to events, for this namespace,
+     respectively.
+     - parameter conflictHandler: a ConflictHandler to handle conflicts on this namespace
+     - parameter changeEventListener: a ChangeEventListener to listen to events on this namespace
+     */
+    mutating func configure<T: ConflictHandler, V: ChangeEventListener>(conflictHandler: T,
```

Comment:
[api discussion] In Java, this pattern works nicely, because you can pass an implementation of an interface inline (or use a lambda). In Swift, you can't really do that AFAIK. I know this isn't where we're going to expose configuration for the user, but have we thought about having configure variants that accept closures instead of protocols?

Reply:
I think it's a good idea to do both. We can store it either way as their type erased equivalents, and the user can go about however they place. IMO, in an actual application, there's a stronger use case for having a singleton conform to these protocols than just a throwaway lambda. However, for the sake of the user, probably best to support both.

Reply:
This will be exposed at a higher level.